crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -153,31 +154,58 @@ impl BlobVec {
-        // collection's `Drop` impl will call `drop` again for the old value (which is still stored
-        // in the collection), so we get a double drop. To prevent that, we set len to 0 until we're
-        // done.
-        let old_len = self.len;
-        let ptr = self.get_unchecked_mut(index).promote().as_ptr();
-        self.len = 0;
-        // Drop the old value, then write back, justifying the promotion
-        // If the drop impl for the old value panics then we run the drop impl for `value` too.
-        if let Some(drop) = self.drop {
-            struct OnDrop<F: FnMut()>(F);
-            impl<F: FnMut()> Drop for OnDrop<F> {
-                fn drop(&mut self) {
-                    (self.0)();
-                }
-            }
-            let value = value.as_ptr();
-            let on_unwind = OnDrop(|| (drop)(OwningPtr::new(NonNull::new_unchecked(value))));
 
-            (drop)(OwningPtr::new(NonNull::new_unchecked(ptr)));
+        let size = self.item_layout.size();
+
+        // Pointer to the value in the vector that will get replaced.
+        let dst;
+        let src = value.as_ptr();
 
+        if let Some(drop) = self.drop {
+            // Temporarily set the length to zero, so that uninitialized bytes won't
+            // get observed in case a panic causes this function to exit early.
+            let old_len = self.len;
+            self.len = 0;
+
+            // Transfer ownership of the old value out of the vector, so it can be dropped.
+            // SAFETY:
+            // - The caller ensures that `index` fits in this vector.
+            // - `size` must be a multiple of the erased type's alignment,
+            //   so adding a multiple of `size` will preserve alignment.
+            // - The storage location will get overwritten with `value` later, which ensures
+            //   that the element will not get observed or double dropped later.
+            // - If a panic occurs, `self.len` will remain `0`, which ensures a double-drop
+            //   does not occur. Instead, all elements will be forgotten.
+            let old_value = self.get_ptr_mut().byte_add(index * size).promote();
+            dst = old_value.as_ptr();
+
+            // This closusure will run in case `drop()` panics, which is similar to
+            // the `try ... catch` construct in languages like C++.
+            // This ensures that `value` does not get forgotten in case of a panic.
+            let on_unwind = OnDrop::new(|| drop(value));
+
+            drop(old_value);
+
+            // If the above code does not panic, make sure that `value` doesn't get dropped.
             core::mem::forget(on_unwind);
+
+            // Make the vector's contents observable again, since panics are no longer possible.
+            self.len = old_len;
+        } else {
+            // SAFETY:
+            // - The caller ensures that `index` fits in this vector.
+            dst = self.get_ptr_mut().as_ptr().add(index * size);
```

Comment:
I dont know why this was moved into here and the `if` branch. This seems strictly worse to me and I'm not clear on what the benefits were supposed to be. Now instead of having a line or two of logic right next to each other that makes it incredibly easy to tell what value `dst` has, you have to go looking all over this function to find out, and also we now are writing the same code twice for fetching the element.

Reply:
This is from this comment thread: https://github.com/bevyengine/bevy/pull/7181#discussion_r1070090478.

We could move this to the line where `dst` is declared and just turn it into a `OwningPtr` in the if block.

Reply:
I'd need to erase the lifetimes in order to pull this line out of the loop. I could do that, but the current way seemed cleaner to me.

Reply:
I was able to make the lifetime erasure method a lot cleaner by defining a `OwningPtr` -> `NonNull` conversion.